### DISCUSSION TITLE
Does segment_UTF8 function accepts offset as an argument?

### DISCUSSION THREAD

**Discussion by riajain0412** [unicode-org/icu4x#3231] [Q&A]
I want to use segmenter functionality with offset. In ICU4C, we have an option to get boundary according to offset so do we have this on the similar lines in ICU4X?
I checked ICU4X segmenter docs but was unable to see if we can pass a offset argument or not.

**Comment by sffc** [comment]
ICU4X should support everything in ECMA-402, which has a `Segments.containing` function:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Segmenter/segment/Segments/containing

```javascript
const segmenterFr = new Intl.Segmenter('fr', { granularity: 'word' });
const string1 = 'Que ma joie demeure';

const segments = segmenterFr.segment(string1);

console.log(segments.containing(5));
// Expected output:
// Object {segment: 'ma', index: 4, input: 'Que ma joie demeure', isWordLike: true}
```

Do we support this yet in ICU4X @makotokato?

**Reply by riajain0412** [reply]
@makotokato , Can you please help me here?

**Reply by sffc** [reply]
CC @aethanyc: is this on the roadmap?

**Reply by riajain0412** [reply]
@aethanyc, Do ICU4X supports segmentation with offset value? If not, then do you guys have this in your roadmap?

**Reply by aethanyc** [reply]
> In ICU4C, we have an option to get boundary according to offset so do we have this on the similar lines in ICU4X?

Are you talking about [ubrk_preceding](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ubrk_8h.html#a00b11da520120f32c0e469e69bf36914) or [ubrk_following](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/ubrk_8h.html#a2b4bf12879b20d7094da3ced996f492f) in ICU4C that sets the iterator to a specified offset? Currently, ICU4X segmenter doesn't have these APIs. We currently also don't support ECMA-402 `Segments` nor `Segments.containing` function. We've not discussed supporting them in the roadmap yet.

However, we might be able to pass a substring into `segmenter_str` to simulate a offset. For example,

```
    let offset = 5;
    let breakpoints = segmenter
        .segment_str(&str[offset..])
        .map(|x| x + offset)
        .collect::<Vec<usize>>();
```

**Reply by aethanyc** [reply]
Filed https://github.com/unicode-org/icu4x/issues/3247.

**Reply by riajain0412** [reply]
Yes, I'm talking about this preceding and following function.

**Reply by riajain0412** [reply]
@sffc , Will ICU4X Test Data provider give correct results for Lao language? 
I was running segment_utf16 on Lao string but its results are not inline with ICU4C results.

**Reply by sffc** [reply]
The ICU4X Test Data provider supports Japanese and Thai. For the other languages, you should follow the steps in the tutorial to generate your own data; in general the testdata provider is intended for testing. You can also track https://github.com/unicode-org/icu4x/issues/2945 which will make it possible to get full data without needing to build it using the tool.

**Reply by riajain0412** [reply]
Okay, is there any tutorial to know how to load data blob in C++?
